### ISSUE TITLE
Make Table accessible

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#10830]
Tables are not very accessible. A simple example can be seen in the "recipe details" screen if the Pesto demo app: In the table of ingredients at the bottom you should be able to select an entire row        when in accessibility node. Currently, you can only select one cell at a time, making accessibility navigation awkward.

**Comment by Hixie** [comment]
This is on our radar for the coming months.

**Comment by goderbauer** [comment]
Another interesting use case is a `Table` containing multiple `Text` widgets on a `Card` widget: currently, all the text from the table is folded into the semantics node for the `Card` and developers would have to manually wrap their cell content into `Semantics` widget, if they want those to be able to get individual focus.

Is that expected? Or should a table by default provide one SemanticsNode per Cell? Per Row? Per Column? Or per Table?

**Comment by Hixie** [comment]
Table probably shouldn't but DataTable probably should.

**Comment by goderbauer** [comment]
From talking to various people over the last few days it seems like we need to offer developers a choice in how the semantics of their table are presented. You should be able to configure that

* each cell is its own semantic nodes,
* each row is one semantics node,
* each column is one semantics node, or
* the table is transparent (just like the Row or Column widget) and doesn't have any opinion about semantics.

**Comment by Hixie** [comment]
cc @jonahwilliams

**Comment by jonahwilliams** [comment]
I can start on this by investigating some existing tables semantic behavior.  @goderbauer , if you have any notes from your research into this it would help greatly.

**Comment by jonahwilliams** [comment]
From offline discussion it seems like there are two major tasks here:

1. Make the data table fully accessible, according to the existing material implementations 

and

2. Make the table semantics configurable by adjusting the render objects so that (1) can be completed.  Due to the nature of the table layout, users cannot just wrap cells/rows in semantics objects.

**Comment by jonahwilliams** [comment]
I am delaying finishing the implementation of this for higher priority work

**Comment by jonahwilliams** [comment]
Dumping my thoughts here since I was unable to get something working this quarter:

The challenge is that changing how semantics are grouped can only currently be implemented in the semantics merging code, not by specific widgets.

In order to make each row selectable, we would need to insert new semantics nodes and them merge them with the correct cells. This does not appear to be supported, for example:

```dart
  @override
  void describeSemanticsConfiguration(SemanticsConfiguration config) {
    super.describeSemanticsConfiguration(config);
    config.isSemanticBoundary = true;
    config.explicitChildNodes = true;
  }


  @override
  void assembleSemanticsNode(SemanticsNode node, SemanticsConfiguration config, Iterable<SemanticsNode> children) {
    final List<SemanticsNode> newChildren = <SemanticsNode>[];
    final List<Rect> rowBoxes = <Rect>[];
    for (int i = 0; i < rows; i++) {
      final Rect rect = getRowBox(i);
      final SemanticsNode child = new SemanticsNode();
      child.rect = rect;
      rowBoxes.add(rect);
      newChildren.add(child);
    }
    int row = 0;
    List<SemanticsNode> rowChildren = <SemanticsNode>[];
    for (SemanticsNode child in children) {
      final Rect currentRow = rowBoxes[row];
      final Vector3 translation = child.transform.getTranslation();
      final Offset offset = new Offset(translation[0], translation[1]);
      print(offset);
      if (currentRow.contains(child.rect.center + offset)) {
        rowChildren.add(child);
      } else {
        final SemanticsConfiguration config = new SemanticsConfiguration()
          ..isSemanticBoundary = true
          ..isMergingSemanticsOfDescendants = true;
        newChildren[row].updateWith(config: config, childrenInInversePaintOrder: rowChildren);
        row++;
        rowChildren = <SemanticsNode>[];
      }
    }
    if (rowChildren.isNotEmpty) {
      final SemanticsConfiguration config = new SemanticsConfiguration()
        ..isSemanticBoundary = true
        ..isMergingSemanticsOfDescendants = true;
      newChildren[row].updateWith(config: config, childrenInInversePaintOrder: rowChildren);
    }
    node.updateWith(config: config, childrenInInversePaintOrder: newChildren);
  }
```

This will immediately throw:

flutter: ══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
flutter: The following assertion was thrown during a scheduler callback:
flutter: 'package:flutter/src/semantics/semantics.dart': Failed assertion: line 2271 pos 16: 'node.parent ==
flutter: null || !node.parent.isPartOfNodeMerging || node.isMergedIntoParent': is not true.

Since the nodes we are inserting are not parents until after they have been updated. It seems the merging logic cannot correctly handle inserting new merging nodes between a tree of existing nodes.

cc @yjbanov @goderbauer

**Comment by yjbanov** [comment]
Inserting nodes in a mutable tree is hard due to the amount of bookkeeping that it requires (imagine inserting HTML DOM nodes where your templates do not have them and managing that). I think it would have been much simpler if semantics tree were reactive, like the widget tree is. Then we would be able to express a many-to-many relationship between render objects and semantics nodes, just like we have a many-to-many relationship between widgets and render objects.

**Comment by laureljanej** [comment]
@jonahwilliams We're using the DataTable in an app that we're doing significant accessibility testing on and have some ideas for enhancements that would make the tables more accessible to screen reader users. These enhancements are generally based on best practices for building accessible tables in HTML but the same concepts apply.

- Add a semantics label to the DataTable that describes the number of columns and rows
- Add a semantics label to each DataCell that associates it with its column header

These labels help non-sighted users and users with low vision to understand the information presented in the table without the visual indication of the column headers. This [article from W3C](https://www.w3.org/WAI/tutorials/tables/) gives a good overview of the desired behavior.

**Comment by mateusfccp** [comment]
We are adding semantics to our app, and we have to insert semantics in a `TableRow`. A `Table` accepts only `List<TableRow>`, so we can't wrap each one in a `Semantics` widget.

We could wrap each cell, but then the reading will be subpar.
